docs/proposals/28-autonomous-shoot-clusters.md

```diff
@@ -0,0 +1,495 @@
+- [`gardenadm version`](#gardenadm-version) to print the `gardenadm` version.
+
+The general flow for bootstrapping an autonomous shoot cluster will be as follows:
+
+1. The user installs `gardenadm` on the first control plane node and runs `gardenadm discover`.
+   This will download relevant configuration from the Gardener system the autonomous shoot cluster shall be later connected to (e.g., `Controller{Registration,Deployment}`s, `CloudProfile`s, etc.).
+   If there is no existing Gardener system yet, the needed configuration must be provided manually.
+2. The downloaded configuration and a `Shoot` manifest are provided to the `gardenadm init` command.
+   This will bootstrap the first control plane node.
+3. Optionally, additional control plane nodes can be joined.
+   In order to do so, the user runs `gardenadm token generate` which generates a random bootstrap token, and then `gardenadm token create <token>` which creates it on the server.
+   Now connected to the new node that shall be joined, the user runs the `gardenadm join` command which bootstraps another control plane node.
+5. Optionally, worker nodes can be joined.
+   The flow is the same - a bootstrap token must be generated and created, and then `gardenadm join` must be executed to bootstrap the worker node.
+6. Once above steps are completed, the user runs `gardenadm connect` to deploy a `gardenlet` into the cluster which connects to an existing Gardener system for further cluster management.
+
+A more detailed description of the commands is following.
+
+#### `gardenadm discover`
+
+This command takes a kubeconfig to an existing Gardener system, and a `Shoot` manifest.
+It downloads the required `CloudProfile`, `ControllerRegistration` and `ControllerDeployment` resources and stores them locally for usage of other `gardenadm` commands.
+This simplifies the crafting of these documents, and it also ensures that the configuration and versions used for the bootstrapping of the autonomous shoot cluster fit to the configuration later found in the connected Gardener system.
+
+#### `gardenadm init`
+
+This command bootstraps the first control plane node of the autonomous shoot cluster.
+The configuration is specified via separate files.
+It needs a `Shoot` manifest and some additional configuration files, which usually reside in the garden cluster, e.g., a `CloudProfile` (best downloaded with `gardenadm discover`).
+For the initial cluster Gardener bootstrapping use-case however, these won't come from a pre-existing garden cluster as it doesn't exist yet.
+They can then be provided locally (that's why `gardenadm discover` and `gardenadm init` are separate commands), reflecting the to-be-deployed Gardener system later.
+
+The high-level steps executed by the command are as follows:
+
+1. Generate certificate authorities, certificates and keys using the [secrets manager](../development/secrets_management.md).
```

Comment:
Secrets manager is now using k8s API server to store the secrets, but in this case there might not be such cluster, what kind of storage will be used?

Reply:
We propose to fake the API server (see https://github.com/gardener/gardener/blob/68ccf134a903568bb724f0926d5c3dfa2e470379/docs/proposals/28-autonomous-shoot-clusters.md#faking-a-kubernetes-api).

Reply:
In the regular seed/shoot setup, we store the secrets in the shoot namespace in the seed cluster. For this autonomous setup, it is just natural to store the secrets directly in the cluster. This way, regular pods like GRM or `kube-proxy` can easily mount them (e.g., CA bundles). There is no other place to store them, or do you have something in mind?

Reply:
Agree 👍.

I was more referring to the fake client that is first used to generate the secrets and the data being present in-memory only. There is no explicit step that causes to save these secrets in the cluster. Thus, I wondered if it was intended this way or if it'll happen differently.

Reply:
Okay, good. Yes, the whole process involves a lot of details, and we refrained from documenting all of them to not overload the content. :)

Reply:
Maybe the fake api server can sync them in files on the node filesystem to not lose them on restart, what do you think?

Reply:
I think it doesn't matter much if they are lost - a new invocation of `gardenadm init` would simply create new secrets. If a previous invocation has reached the point where they have been replicated into the cluster as `Secret`s, they will be picked up from there and not regenerated. This way, we don't need to implement our own fake client and can simply use the existing one.

Reply:
Yes, something like this (that's at least how we've done it in the PoC).